### PR TITLE
SNOW-500400: Allow structured data to be literal value 

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-500400

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

    Make list, tuple and dict a valid literal value in snowpark

### CODE REVIEW

File: src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py [hunk 1]
```diff
@@ -57,31 +53,19 @@ def to_sql(value: Any, datatype: DataType, from_values_statement: bool = False)
             return "NULL"
     if isinstance(datatype, BinaryType):
         if value is None:
-            return "NULL :: binary"
-    if isinstance(datatype, IntegerType):
+            return "NULL :: BINARY"
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py:56]: Changes here are nor really related to literal value - it's just to make translated sql more formal (https://docs.snowflake.com/en/sql-reference/data-types-numeric.html#int-integer-bigint-smallint-tinyint-byteint, we can just use INT and FLOAT) and readable (uppercase).

File: src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py [hunk 2]
```diff
@@ -155,13 +126,13 @@ def to_sql(value: Any, datatype: DataType, from_values_statement: bool = False)
             return f"TIME('{trimmed_ms}')"
 
     if isinstance(value, (list, bytes, bytearray)) and isinstance(datatype, BinaryType):
-        return f"'{binascii.hexlify(value).decode()}' :: binary"
+        return f"'{binascii.hexlify(value).decode()}' :: BINARY"
 
     if isinstance(value, (list, tuple, array)) and isinstance(datatype, ArrayType):
-        return f"parse_json({str_to_sql(json.dumps(value))})"
+        return f"PARSE_JSON({str_to_sql(json.dumps(value))}) :: ARRAY"
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py:132]: These two changes here are needed - otherwise the data type will be variant.

**Review by sfc-gh-mabrennan — APPROVED** [review]
Docstrings look great.

File: src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py [hunk 3]
```diff
@@ -91,41 +75,28 @@ def to_sql(value: Any, datatype: DataType, from_values_statement: bool = False)
...
+    if isinstance(value, float) and isinstance(datatype, _FractionalType):
         if math.isnan(value):
-            cast_value = "'Nan'"
+            cast_value = "'NaN'"
         elif math.isinf(value) and value > 0:
-            cast_value = "'Infinity'"
+            cast_value = "'inf'"
         elif math.isinf(value) and value < 0:
-            cast_value = "'-Infinity'"
+            cast_value = "'-inf'"
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py:95]: Curious: why do we change this?

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py (hunk 3 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py:95]: just make the sql more formal, which are suggested by snowflake doc https://docs.snowflake.com/en/sql-reference/data-types-numeric.html#special-values

**Pull request merged**